src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,399 @@
+   `pedantic`, `test`, and `examples_util`
+   (as listed in the **Resolvable** column).
+3. Run `dart pub upgrade`.
+4. Run `dart pub outdated --mode=null-safety` once more,
+   to make sure that all dependencies are null safe.
+
+{{ site.alert.info }}
+  **Why do all dependencies need to support null safety?**
+  When all direct dependencies support null safety,
+  you can _run your app_ with sound null safety.
+  When all dev dependencies support null safety,
+  you can _run tests_ with sound null safety.
+{{ site.alert.end }}
+
+
+## 2. Migrate {#step2-migrate}
+
+Most of the changes that your code needs to be null safe
+are easily predictable.
+For example, if a variable can be `null`,
+its type needs a `?` suffix.
+A named parameter that shouldn't be nullable
+needs to be marked `required`.
+
+The migration tool can make most of these changes,
+and you can add hints to customize the changes.
+Alternatively, you can
+[migrate your code by hand](#migrating-by-hand).
+
+{{ site.alert.tip }}
+  For additional help while migrating code, check the
+  [null safety FAQ][].
+{{ site.alert.end }}
+
+[null safety FAQ]: /null-safety/faq
+
+
+### Using the migration tool
+
+The migration tool takes a package of null-unsafe Dart code
+and converts it to null safety.
+
+Before running the tool,
+make sure you're using a 2.12 beta release (or later):
+
+```terminal
+$ dart --version
+Dart SDK version: 2.12.0-18.0.beta (beta)... [PENDING: check version #/string!]
+```
+
+Then update your `pubspec.yaml` file:
+
+* **Keep any null-safe package versions** you chose when you
+  [checked dependency status](#check-dependency-status) in step 1.
+* If your package's SDK constraints have a minimum version of 2.12 or higher:
+  1. **Set the minimum SDK version** back to a version before null safety.
+     ```yaml
+environment:
+    sdk: '>=2.9.0 <3.0.0'
+```
+  2. **Update dependencies:**
+     ```terminal
+$ dart pub get
+```
+
+Next, start the migration tool by running the `dart migrate` command
+in the directory that contains the package's `pubspec.yaml` file:
+
+```terminal
+$ dart migrate
+```
+
+If your package is ready to migrate,
+then the tool produces a line like the following:
+
+```terminal
+View the migration suggestions by visiting:
+
+  http://127.0.0.1:60278/Users/you/project/mypkg.console-simple?authToken=Xfz0jvpyeMI%3D
+```
+
+Visit that URL in a Chrome browser
+to see an interactive UI
+where you can guide the migration process:
+
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+
+For every variable and type annotation,
+you can see what nullability the tool infers.
+For example, in the preceding screenshot,
+the tool infers that the `bar2()` function returns a non-nullable type (`int`).
+
+{{ site.alert.tip }}
+  To see the reason for each change (or non-change),
+  click the type.
+{{ site.alert.end }}
+
+When analysis infers the wrong nullability,
+you can override its choice by inserting temporary `/*?*/` and `/*!*/` markers.
```

Comment:
I think we should consider elaborating here. It's not that clear when to add these hints. The readme has a nice example, I think:
https://github.com/dart-lang/sdk/tree/master/pkg/nnbd_migration#interactive-feedback-to-the-tool

Reply:
I've elaborated the heck out of this. Testing the readme's example, I found that it had some nonstandard code and the results were different, so I used my own versions of those. ptal @mit-mit.

Reply:
Love it, thanks much!